src/interrupts.rs

```diff
@@ -0,0 +1,305 @@
+use std::mem;
+use std::os::unix::io::RawFd;
+use std::collections::VecDeque;
+use epoll::Event;
+use epoll::Events;
+use std::time::Instant;
+use libc;
+use std::error::Error;
+
+pub const MOVING_AVERAGE_RANGE: usize = 5;
+const INTERRUPT_THRESHOLD: u64 = 1_200;
+pub const INTERRUPT_INITIAL_INTERVAL: u64 = 1_000_000_000;
+const MAX_INTERRUPT_VECTORS: u32 = 32;
+
+#[derive(Default)]
+pub struct Interrupts {
+    pub interrupts_enabled: bool, // Whether interrupts for this device are enabled or disabled.
+    pub itr_rate: u32, // The Interrupt Throttling Rate
+    pub interrupt_type: u64, // MSI or MSIX
+    pub timeout_ms: i16, // interrupt timeout in milliseconds (-1 to disable the timeout)
+    pub queues: Vec<InterruptsQueue>,  // Interrupt settings per queue
+}
+
+pub struct InterruptsQueue {
+    pub vfio_event_fd: RawFd, // event fd
+    pub vfio_epoll_fd: RawFd, // epoll fd
+    pub interrupt_enabled: bool, // Whether interrupt for this queue is enabled or not
+    pub last_time_checked: Instant, // Last time the interrupt flag was checked
+    pub rx_pkts: u64, // The number of received packets since the last check
+    pub interval: u64, // The interval to check the interrupt flag
+    pub moving_avg: InterruptMovingAvg, // The moving average of the hybrid interrupt
+}
+
+#[derive(Default)]
+pub struct InterruptMovingAvg {
+    pub measured_rates: VecDeque<u64>, // The moving average window
+    pub sum: u64, // The moving average sum
+}
+
+/// constants and structs needed for IOMMU. Grabbed from linux/vfio.h
+/// as this is a dynamically sized struct (has an array at the end) we need to use
+/// Dynamically Sized Types (DSTs) which can be found at
+/// https://doc.rust-lang.org/nomicon/exotic-sizes.html#dynamically-sized-types-dsts
+#[allow(non_camel_case_types)]
+#[repr(C)]
+struct vfio_irq_set<T: ?Sized> {
+    argsz: u32,
+    flags: u32,
+    index: u32,
+    start: u32,
+    count: u32,
+    data: T,
+}
+
+#[allow(non_camel_case_types)]
+#[repr(C)]
+struct vfio_irq_info {
+    argsz: u32,
+    flags: u32,
+    index: u32,		/* IRQ index */
+    count: u32,		/* Number of IRQs within this index */
+}
+
+const VFIO_IRQ_SET_DATA_NONE: u32 = (1 << 0); /* Data not present */
+const VFIO_IRQ_SET_DATA_EVENTFD: u32 = (1 << 2); /* Data is eventfd (s32) */
+const VFIO_IRQ_SET_ACTION_TRIGGER: u32 = (1 << 5); /* Trigger interrupt */
+const VFIO_DEVICE_GET_IRQ_INFO: u64 = 15213;
+const VFIO_DEVICE_SET_IRQS: u64 = 15214;
+pub const VFIO_PCI_MSI_IRQ_INDEX: u64 = 1;
+pub const VFIO_PCI_MSIX_IRQ_INDEX: u64 = 2;
+const VFIO_IRQ_INFO_EVENTFD: u32 = (1 << 0);
```

Comment:
These structs and constants might fit better into `vfio.rs` to have all vfio related structs and constants in one file?

Btw: This is one of the reasons you should rebase :D There is no `vfio.rs` in your branch yet